### ISSUE TITLE
Nullable reference type checking ignores [Optional][DefaultParameterValue(null)]

### ISSUE THREAD

**Issue by voroninp** [dotnet/csharplang#3268]
With enabled nullable types:

```CSharp
#nullable enable
void Foo(object param = null) {}
```
I get the warning from the compiler: > Cannot convert `null` literal to non-nullable reference type.
However, if I change signature to this:

```CSharp
#nullable enable
void Foo([Optional, DefaultParameterValue(null)] object param ) {}
```
Compiler is totally fine with this. 

P.S. I've discovered this issue for the case of caller member name:
```CSharp
[CallerMemberName, DisallowNull, Optional, DefaultParameterValue(null)] string memberName
```

**Comment by Joe4evr** [comment]
You really shouldn't be using the attributes `[Optional]` and `[DefaultParameterValue]` manually in the first place.

**Comment by voroninp** [comment]
There are many things which I shouldn't do, but which are fine for the compiler.

**Comment by jnm2** [comment]
`[CallerMemberName] string memberName = null!` is preferable. It emits different IL because DisallowNull is not present (it would be redundant).

**Comment by voroninp** [comment]
@jnm2 , I agree, but `[DisallowNull]` would indicate to the caller that `null` is not ok. Anyway, this parameter should not be used explicitly, but it is still possible.

**Comment by alrz** [comment]
>Anyway, this parameter should not be used explicitly, but it is still possible.

R# does warn on such usages. I think this is a good candidate for warning waves.

**Comment by jnm2** [comment]
@voroninp DisallowNull is redundant because the parameter type `string` is already non-nullable. DisallowNull was designed to be applied to generic types or to nullable reference types where the expression must not be null before the call but may be null when the call returns. Because this is not a `ref` parameter or a generic type, DisallowNull is the wrong thing to use.

**Comment by AartBluestoke** [comment]
> Because this is not a `ref` parameter or a generic type, DisallowNull is the wrong thing to use.

Sounds like this should be a warning then :)

**Comment by jnm2** [comment]
I agree, and I wish analyzers were considered in every API pass. Lots of untapped timesavers hiding around the BCL IMO.
